include/alpaka/core/Concepts.hpp

```diff
@@ -0,0 +1,53 @@
+/* Copyright 2019 Benjamin Worpitz
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <type_traits>
+
+namespace alpaka
+{
+    namespace concepts
+    {
+        //#############################################################################
+        //! Tag used in class inheritance hierarchies that describes that a specific concept (TConcept)
+        //! is implemented by the given base class (TBase).
+        template<
+            typename TConcept,
+            typename TBase>
```

Comment:
I am a little confused, why base class here? It looks like the type actually provided here is a child class that implements the concept and inherits from `Implements`

Reply:
Before this change we had inheritance hierarchies like `AccYyy->XxxImpl` where `XxxImpl` had a typedef that tagged it `using XxxBase = XxxImpl;` as implementing the Xxx concept.
This PR transformed it into `AccYyy->XxxImpl->Implements<ConceptXxx, XxxImpl>`.
So the `XxxBase` typedef has been transformed into the `TBase` template parameter. Therefore the `Base` name ;-)
The inheritance instead of the tag will allow us in the future to check at compile time that a specific class implements a concept only once which is not possible with typedefs.
Currently the implementation classes tag themselves as `XxxImpl->Implements<ConceptXxx, XxxImpl>`. This was the same for the typedef and in this case it does not really look like a `Base` class that is given as a parameter. Looking at it from a language standpoint is is not wrong either because `std::is_base_of<T, T>::value` is always true ;-)
However, I am thinking about moving the tagging onto the most derived class at least for some cases which would then look like this: `AccYyyl->Implements<ConceptXxx, XxxImpl>`. In this case the parameter really is a strict base-class. My use case would be the `AtomicHierarchy`. The atomic implementations can not know for which atomic hierarchy they are used (grid, block, thread) so they can not tag themselves as implementing a specific concept. The `AtomicHierarchy` on the other hand can tag itself and say that Impl1 implements the grid and block atomics and Impl2 implements the thread atomics.